File: JavaFactorial/src/de/luschny/math/factorial/FactorialFactors.java
// Copyright (C) 2004-2009 Peter Luschny, MIT License applies.
// See http://en.wikipedia.org/wiki/MIT_License
// Visit http://www.luschny.de/math/factorial/FastFactorialFunctions.htm
// Comments mail to: peter(at)luschny.de
package de.luschny.math.factorial;

import de.luschny.math.primes.IPrimeIteration;
import de.luschny.math.primes.IPrimeSieve;
import de.luschny.math.primes.PrimeSieve;

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;

public class FactorialFactors {

    public FactorialFactors() {
    }
    private int[] primeList;
    private int[] multiList;
    private int n, card;

    public FactorialFactors(int n) {
        this.n = n;
        card = primeFactors(n);
    }

    public void saveToFile(String fileName) {
        try {
            PrintWriter primeReport = new PrintWriter(new BufferedWriter(new FileWriter(fileName)));

            writeFactors(primeReport);
            primeReport.close();
        } catch (IOException e) {
            System.err.println(e.toString());
        }
    }

    public void writeFactors(PrintWriter file) {
        file.println("The prime factors of " + this.n + "! ");

        int sum = n - Integer.bitCount(n);
        file.print("2^" + sum);

        for (int p = 0; p < this.card; p++) {
            int f = primeList[p], m = multiList[p];
            sum += m;

            if (m > 1) {
                file.print("*" + f + "^" + m);
            } else {
                file.print("*" + f);
            }
        }

        file.println();
        file.println("Number of different factors: " + this.card);
        file.println("Number of all factors: " + sum);
    }

    private int primeFactors(int k) {
        IPrimeSieve sieve = new PrimeSieve(k);
        IPrimeIteration pIter = sieve.getIteration(2, k);

        int piN = pIter.getNumberOfPrimes();

        primeList = new int[piN];
        multiList = new int[piN];

        int maxBound = k / 2, count = 0;

        for (int prime : pIter) {
            int m = prime > maxBound ? 1 : 0;

            if (prime <= maxBound) {
                int q = k;
                while (q >= prime) {
                    m += q /= prime;
                }
            }

            primeList[count] = prime;
            multiList[count++] = m;
        }
        return count;
    }
}

File: SilverFactorial64/Sharith/Factorial/FactorialFactors.cs
// -------- ToujoursEnBeta
// Author & Copyright : Peter Luschny
// License: LGPL version 3.0 or (at your option)
// Creative Commons Attribution-ShareAlike 3.0
// Comments mail to: peter(at)luschny.de
// Created: 2010-03-01

namespace Sharith.Factorial
{
    using System.IO;
    using Sharith.Primes;
    using XMath = MathUtils.XMath;

    public class FactorialFactors
    {
        public FactorialFactors()
        {
        }

        public FactorialFactors(int n)
        {
            this.Factor(n);
        }

        int[] primeList;
        int[] multiList;
        int n, card;

        public int Factor(int n)
        {
            this.n = n;

            var lgN = XMath.FloorLog2(n);
            var piN = 2 + (15 * n) / (8 * (lgN - 1));

            this.primeList = new int[piN];
            this.multiList = new int[piN];

            return this.card = this.PrimeFactors(n);
        }

        public void SaveToFile(string fileName)
        {
            var file = new FileInfo(@fileName);
            var stream = file.AppendText();
            this.WriteFactors(stream);
            stream.Close();
        }

        public void WriteFactors(TextWriter file)
        {
            if (file == null)
            {
                return;
            }
            file.WriteLine("The prime factors of {0}! ", this.n);

            var sum = this.n - XMath.BitCount(this.n);
            file.Write("2^{0}", sum);

            for (var p = 0; p < this.card; p++)
            {
                int f = this.primeList[p], m = this.multiList[p];
                sum += m;
                if (m > 1)
                {
                    file.Write("*{0}^{1}", f, m);
                }
                else
                {
                    file.Write("*{0}", f);
                }
            }

            file.WriteLine();
            file.WriteLine("Number of different factors: {0} ", this.card);
            file.WriteLine("Number of all factors:       {0} ", sum);
        }

        private int PrimeFactors(int n)
        {
            var sieve = new PrimeSieve(n);
            var primeCollection = sieve.GetPrimeCollection(2, n);

            int maxBound = n / 2, count = 0;

            foreach (var prime in primeCollection)
            {
                var m = prime > maxBound ? 1 : 0;

                if (prime <= maxBound)
                {
                    var q = n;
                    while (q >= prime)
                    {
                        m += q /= prime;
                    }
                }

                this.primeList[count] = prime;
                this.multiList[count++] = m;
            }

            return count;
        }
    }
}
